provider/k8s/metric_scraper.go

```diff
@@ -0,0 +1,144 @@
+package k8s
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/convox/convox/pkg/structs"
+	"github.com/pkg/errors"
+	am "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type MetricScraperClient struct {
+	c       *http.Client
+	cluster kubernetes.Interface
+	host    string
+}
+
+func NewMetricScraperClient(cluster kubernetes.Interface, host string) *MetricScraperClient {
+	return &MetricScraperClient{
+		c: &http.Client{
+			Timeout: 10 * time.Second,
+		},
+		cluster: cluster,
+		host:    host,
+	}
+}
+
+func (m *MetricScraperClient) GetRackMetrics(opts structs.MetricsOptions) (structs.Metrics, error) {
+	if m.host == "" {
+		return nil, errors.WithStack(fmt.Errorf("unimplemented"))
+	}
+
+	ns, err := m.cluster.CoreV1().Nodes().List(am.ListOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	nodeNames := []string{}
+	var cpuAllocatable, memAllocatable float64
+	for _, n := range ns.Items {
+		nodeNames = append(nodeNames, n.ObjectMeta.Name)
+		cpuAllocatable += float64(n.Status.Allocatable.Cpu().MilliValue())
+		memAllocatable += float64(n.Status.Allocatable.Memory().Value())
+	}
+
+	cpus, err := m.GetNodesMetrics(strings.Join(nodeNames, ","), structs.ScraperMetricTypeCpu)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	if len(cpus.Items) == 0 {
+		return nil, fmt.Errorf("metrics unavailable")
+	}
+
+	for i := 1; i < len(cpus.Items); i++ {
+		for j, d := range cpus.Items[i].MetricPoints {
+			if j < len(cpus.Items[0].MetricPoints) {
```

Comment:
in v3 rack means the eks cluster. cluster is consists of nodes. `metrics_scrapper` collects the nodes metrics and in the http response it sends all the node metrics, so summation of these all metrics represents the rack metrics and that's why adding them to position 0. each item in the items[] represents a single node metrics.

 Normally all the nodes will have same amount of data points. But it's may be possible some of them have may be less or more data (the diff will not cross 1) depending on the time we are making the request, that's why added idx check with MetricsPoints at 0. so that there will be no out of index panic

Reply:
ok, that make sense. thanks.
do you think we could try to make it a bit more easier to read?
i thought something like

```go
// aggregate all metric points into mps
mps := cpus.Items[0].MetricPoints

for i := 1; i < len(cpus.Items); i++ {
	for j, mp := range cpus.Items[i].MetricPoints {
		// only consider items with the same number of MetricPoints
		if j < len(mps) {
			    mps[j].Value += mp.Value
		}
	}
}
```

it doesn't need to be precisely that. the idea is to make it a bit easier to read(with comments or renaming/creating new variables)